### DISCUSSION TITLE
git mode for lib microservice

### DISCUSSION THREAD

**Discussion by prasadtalasila** [INTO-CPS-Association/DTaaS#951] [Ideas]
The libms currently supports **local file system mode**. A new feature is to support **git mode**. The requirement is illustrated here.

![image](https://github.com/user-attachments/assets/1df0ac05-01ce-4049-b826-50045817fb91)

The user files are accessible at _/workspace_ and _/workspace/common_. These files need to served in HTTP mode via **cloudcmd**. The files are also accessible at terminal, vscode and Jupyter lab services. The git mode needs to be visible to all services except in HTTP mode.

**Comment by prasadtalasila** [comment]
We used to have gitlab mode in **libms** wherein the gitlab is queried to fetch the files. This gitlab mode became non-functional due to upgrade of gitlab GraphQL API. The PR #435 has removed the gitlab mode. The changes made in that PR can show the scaffolding needed. Please use as an inspiration only; the design can perhaps be improved.

Thus a decision has been made to switch to `git` protocol. Using plain git has the advantage of integration with multiple git providers like gitlab, github, gitolite etc.

**Comment by prasadtalasila** [comment]
@nichlaes Please see the design suggested here.

The [NestJS](https://docs.nestjs.com/) framework is being used to create the lib microservice. The current code only supports local files mode and it has the following structure.

![now](https://github.com/user-attachments/assets/fe7e628f-6050-4751-97bc-eee2f9d01a8e)
The **FileServiceFactory** and **LocalFilesService** are essentially providing the localfiles functionality. The **FilesResolver** is servicing GraphQL queries.

We could refactor code along the lines of **mode** into the following.

![proposed](https://github.com/user-attachments/assets/131c7285-9950-4faf-8b63-ef37243c31a8)

The advantage is the scoping of code into relevant modules. The future development will require adding git commands and management code to git module. The same goes for other proposed modes such as seaweed file system mode.

**Comment by nichlaes** [comment]
![382708804-acac5005-0856-42bf-bb8c-cd1634864092](https://github.com/user-attachments/assets/a4c881c8-57a0-4ae1-816e-6293016f7d15)


The screenshot above, shows the overall architecture for the Libms.

The next approach seems to be a feature for cloning a gitlab repository  (or any other git repo) through isomorphic git.

**Comment by nichlaes** [comment]
@prasadtalasila I think I could benefit, if I had a use case scenario. Like how the user would interact.

Should I make endpoints for getting the `git status` and for commiting?

When should the libms pull? Automatically or should the user do it?

**Reply by prasadtalasila** [reply]
> @prasadtalasila I think I could benefit, if I had a use case scenario. Like how the user would interact.
> 
> Should I make endpoints for getting the `git status` and for commiting?
> 
> When should the libms pull? Automatically or should the user do it?

please see issue #1040
